yarn-project/end-to-end/src/e2e_keys.test.ts

```diff
@@ -0,0 +1,128 @@
+    // a note of funds, when you create the transient you are nullifying that note. So even if I cannot see when you
+    // nullified the transient ones, I can see that you nullified the first.
+    //
+    // E.g.: Say you have a note A, which is 10 $, you nullify it (I can see) and create B and C, that you then spend.
+    // I cannot see B and C, but I saw A, so I knew that you did something with those funds.
+    //
+    //    There are some examples where the action is fully hidden though. One of those examples is shielding where you
+    // instantly consume the note after creating it. In this case, the nullifier is never emitted and hence the action
+    // is impossible to detect with this scheme.
+    //    Another example is a withdraw is withdrawing from DeFi and then immediately spending the funds. In this case,
+    // we would need nsk_app and the contract address of the DeFi contract to detect the nullification of the initial
+    // note.
+    it('nsk_app and contract address are enough to detect note nullification', async () => {
+      const masterNullifierSecretKey = deriveMasterNullifierSecretKey(secret);
+      const nskApp = computeAppNullifierSecretKey(masterNullifierSecretKey, testContract.address);
+
+      const noteValue = 5;
+      const noteOwner = account.getAddress();
+      const noteStorageSlot = 12;
+
+      await testContract.methods.call_create_note(noteValue, noteOwner, noteStorageSlot).send().wait();
+
+      expect(await getNumNullifiedNotes(nskApp, testContract.address)).toEqual(0);
+
+      await testContract.methods.call_destroy_note(noteStorageSlot).send().wait();
+
+      expect(await getNumNullifiedNotes(nskApp, testContract.address)).toEqual(1);
+    });
+
+    const getNumNullifiedNotes = async (nskApp: Fr, contractAddress: AztecAddress) => {
+      // 1. Get all the note hashes
+      const blocks = await aztecNode.getBlocks(INITIAL_L2_BLOCK_NUM, 1000);
+      const noteHashes = blocks.flatMap((block: L2Block) =>
+        block.body.txEffects.flatMap(txEffect => txEffect.noteHashes),
+      );
+      // 2. Get all the seen nullifiers
+      const nullifiers = blocks.flatMap((block: L2Block) =>
+        block.body.txEffects.flatMap(txEffect => txEffect.nullifiers),
+      );
+      // 3. Derive all the possible nullifiers using nskApp
+      const derivedNullifiers = noteHashes.map(noteHash => {
+        const innerNullifier = poseidon2Hash([noteHash, nskApp, GeneratorIndex.NOTE_NULLIFIER]);
+        return siloNullifier(contractAddress, innerNullifier);
+      });
+      // 4. Count the number of derived nullifiers that are in the nullifiers array
+      return derivedNullifiers.reduce((count, derived) => {
+        if (nullifiers.some(nullifier => nullifier.equals(derived))) {
+          count++;
+        }
+        return count;
+      }, 0);
+    };
+  });
+
+  describe('ovsk_app', () => {
+    it('gets ovsk_app', async () => {
```

Comment:
This test case is new. Had to add it since the `request_ovsk_app` is not yet used anywhere so it might be dysfunctional if not checked by this test.

Reply:
![Noice Thats Nice GIF](https://media4.giphy.com/media/yJFeycRK2DB4c/giphy.gif)